scripts/linking.py

```diff
@@ -392,39 +421,62 @@ def eval_candidate_generation(examples: List[data_util.MedMentionExample],
                 else:
                     doc = docs[i]
                     ner_entities = [ent for ent in doc.ents]
+                    predicted_mention_types = [ent.label_ for ent in doc.ents]
                     mention_texts = [ent.text for ent in doc.ents]
 
                 batch_candidate_neighbor_ids = candidate_generator.generate_candidates(mention_texts, k)
 
+                filtered_batch_candidate_neighbor_ids = []
                 for candidate_neighbor_ids in batch_candidate_neighbor_ids:
                     # Keep only canonical entities for which at least one mention has a score less than the threshold.
                     filtered_ids = {k: v for k, v in candidate_neighbor_ids.items() if any([z[1] <= threshold for z in v])}
+                    filtered_batch_candidate_neighbor_ids.append(filtered_ids)
                     num_candidates.append(len(candidate_neighbor_ids))
                     num_filtered_candidates.append(len(filtered_ids))
                     doc_candidates.update(filtered_ids)
 
                 for i, gold_entity in enumerate(entities):
                     if use_gold_mentions:
-                        candidates = batch_candidate_neighbor_ids[i]  # for gold mentions, len(entities) == len(batch_candidate_neighbor_ids)
+                        candidates = filtered_batch_candidate_neighbor_ids[i]  # for gold mentions, len(entities) == len(filtered_batch_candidate_neighbor_ids)
+                        mention_types = umls_concept_dict_by_id[gold_entity.umls_id]['types']  # use gold types
                     else:
                         # for each gold entity, search for a corresponding predicted entity that has the same span
                         span_from_doc = doc.char_span(gold_entity.start, gold_entity.end)
                         candidates = {}
                         for j, predicted_entity in enumerate(ner_entities):
                             if predicted_entity == span_from_doc:
-                                candidates = batch_candidate_neighbor_ids[j]
+                                candidates = filtered_batch_candidate_neighbor_ids[j]
+                                mention_types = predicted_mention_types[j]
                                 break
 
-                    # Keep only canonical entities for which at least one mention has a score less than the threshold.
-                    filtered_ids = {k: v for k, v in candidates.items() if any([z[1] <= threshold for z in v])}
-
-                    if len(filtered_ids) == 0:
+                    # Evaluating candidate generation
+                    if len(candidates) == 0:
                         entity_no_links_count += 1
                     elif gold_entity.umls_id in candidates:
                         entity_correct_links_count += 1
                     else:
                         entity_wrong_links_count += 1
 
+                    # Evaluating linking
```

Comment:
This should be it's own function outside the body of this loop, which does the classification for a single (mention, candidate_list) pair.

Reply:
added a Linker class (https://github.com/allenai/scispacy/pull/106)